File: graph/leetcode_543.java
package graph;

public class leetcode_543 {

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int diameterOfBinaryTree(TreeNode root) {
        return DFS(root)[0];
    }

    private int[] DFS(TreeNode node) {
        if (node == null) return new int[]{0, 0};
        int[] left = DFS(node.left);
        int[] right = DFS(node.right);

        int best = Math.max(left[1] + right[1], Math.max(left[0], right[0]));
        int height = 1 + Math.max(left[1], right[1]);
        return new int[]{best, height};
    }
}

File: tree/LeetCode_543.kt
package tree

import util.TreeNode

class LeetCode_543 {

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        return findDiameter(root).first
    }

    private fun findDiameter(node: TreeNode?): Pair<Int, Int> {
        if (node == null) {
            return Pair(0, 0)
        }

        val left = findDiameter(node.left)
        val right = findDiameter(node.right)

        val longestDiameter = maxOf(
            maxOf(left.first, right.first),
            (left.second + right.second)
        )
        val height = maxOf(left.second, right.second) + 1

        return Pair(longestDiameter, height)
    }
}

fun main() {
    // Input
    val root = TreeNode.of(1, 2, 3, 4, 5)
    // Output
    LeetCode_543().diameterOfBinaryTree(root).run {
        println(this)
        require(3 == this)
    }
}
